### PR TITLE
Add Performance Best Practices document

### LINKED ISSUE #4491: Add document for performance best practices
It is better to have a official guidance to how to maximize the performance.
I'm thinking about covering these topics, but please feel free to comment if you have other ideas.

* Use the latest Chainer
* Use CuPy
* Use iDeep on CPU
* Install cuDNN
* Set cuDNN workspace size
* Enable auto tuning
* Next Steps (brief introduction of ChainerMN)

### PR DESCRIPTION
**Pull request by kmaehashi**
Fixes #4491.

### CODE REVIEW

File: docs/source/performance.rst [hunk 1]
```diff
@@ -0,0 +1,174 @@
+Performance Best Practices
+==========================
+
+This is a guide to maximize the performance of Chainer.
```
Comment (bkvogel) [inline docs/source/performance.rst]: Consider rewording: This is a guide to maximize the performance of Chainer.  ->
This guide explains some tips and advice for maximizing the performance of Chainer.

File: docs/source/performance.rst [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+
+Use the Latest Version
+----------------------
+
+It is generally recommended to use the latest version of Chainer and its dependent libraries (CUDA, cuDNN, iDeep, etc.).
+Part of new features and performance optimizations introduced in newer versions of dependent libraries may not be available in older versions of Chainer.
```
Comment (bkvogel) [inline docs/source/performance.rst]: Part of new features -> Some of the new features

File: docs/source/performance.rst [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+Also, Chainer itself is incrementally being improved to provide better performance.
+
+If you are using Chainer v4 or later, you can check version configuration by:
```
Comment (bkvogel) [inline docs/source/performance.rst]: check version -> check the version

File: docs/source/performance.rst [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+      CuPy Version          : 4.0.0
+      CUDA Root             : /usr/local/cuda
+      CUDA Build Version    : 9000
+      CUDA Driver Version   : 9000
+      CUDA Runtime Version  : 9000
+      cuDNN Build Version   : 7100
+      cuDNN Version         : 7100
+      NCCL Build Version    : 2102
+
+Generally, Chainer team is maintaining the API between minor updates (e.g., v4.0 to v4.1) so that users can upgrade Chainer without modifying the code (see :doc:`compatibility` for our policy).
```
Comment (bkvogel) [inline docs/source/performance.rst]: Generally, Chainer team -> Generally, the Chainer team

Comment (bkvogel) [inline docs/source/performance.rst]: without modifying the code -> without modifying their code

File: docs/source/performance.rst [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    8388608
+
+and can be adjusted using :func:`chainer.backends.cuda.set_max_workspace_size`.
+
+Maximum required workspace size may vary depending on various conditions such as GPU hardware and batch size of inputs.
+
+Auto-Tuner
+~~~~~~~~~~
+
+Some convolution algorithms in cuDNN support auto-tuner feature that finds the fastest convolution algorithm for given inputs.
```
Comment (bkvogel) [inline docs/source/performance.rst]: auto-tuner -> the auto-tuner

File: docs/source/performance.rst [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+You can turn on this feature by setting ``autotune`` configuration to ``True``.
+
+See :doc:`reference/configuration` for detailed descriptions.
+
+.. note::
+
+    Auto-tuner tries to find the best algorithm for every first observation of the input shape combination.
+    Therefore, the first batch will become slower when auto-tuner is enabled.
+    The result of auto-tuner is cached so that it can be reused for data with the same input shape combination.
```
Comment (bkvogel) [inline docs/source/performance.rst]: Will the cached result of the auto-tuner be reused if the model is later run with ``autotuned`` set to ``False``? This is not clear to me from the description. Or should the user then set it to ``False`` for later runs?

File: docs/source/reference/configuration.rst [hunk 7]
```diff
@@ -25,79 +25,116 @@ The following entries of the configuration are currently provided by Chainer.
...
 
-``chainer.config.cudnn_deterministic``
+Configuration Keys
+------------------
+
+* ``cudnn_deterministic`` (default: ``False``)
    Flag to configure deterministic computations in cuDNN APIs.
+
    If it is ``True``, convolution functions that use cuDNN use the deterministic mode (i.e, the computation is reproducible).
    Otherwise, the results of convolution functions using cuDNN may be non-deterministic in exchange for the performance.
```
Comment (bkvogel) [inline docs/source/reference/configuration.rst]: in exchange for the performance -> in exchange for better performance

File: docs/source/reference/configuration.rst [hunk 8]
```diff
@@ -25,79 +25,116 @@ The following entries of the configuration are currently provided by Chainer.
...
-   The default value is ``False``.
-``chainer.config.debug``
+
+* ``debug`` (default: ``False``)
    Debug mode flag.
+
    If it is ``True``, Chainer runs in the debug mode.
```
Comment (bkvogel) [inline docs/source/reference/configuration.rst]: in the debug mode -> in debug mode

**Comment by kmaehashi** [comment]
@bkvogel Thank you for reviewing! Fixed accordingly.

**Comment by bkvogel** [comment]
LGTM!

**Comment by kmaehashi** [comment]
@bkvogel This pull-request is marked as `to-be-backported`, but the corresponding backport PR could not be found. Could you check?

**Comment by kmaehashi** [comment]
I made a backport PR.

**Pull request merged**